### DISCUSSION TITLE
集群消费模式下，是否在某些情况下，存在同一个队列被 同一个消费组的 多个消费者消费的情况？

### DISCUSSION THREAD

**Discussion by bjsfdx2004** [apache/rocketmq#7746] [Q&A]
我在源码里面没有看到对上面问题的严格的代码限制，所以感觉在下面情况下，会出现
同一个队列被 同一个消费组的 多个消费者消费的情况：

1. 比如自定义的 AllocateMessageQueueStrategy 把同一个队列，分配给同一个消费组的 多个消费者
2. 服务端发送的 NOTIFY_CONSUMER_IDS_CHANGED one way 消息，没有被某些消费者收到，所以不知道有新consumer加入

我的理解是上面两种情况，都会出现 同一个队列被 同一个消费组的 多个消费者消费的情况？
PS：我看代码，发现同一个消费组的 多个消费者 各自的 SubscriptionData 的 subVersion 允许不一样，这个字段多个消费者之间没有做同步，是允许不一致吗？

**Comment by cserwen** [comment]
RocketMQ 支持 at-least-once 语义，至少投递一次，不保证只消费一次，然后上述情况是会出现的
- Rebalance 是最终一致，除了服务端主动通知，各个 consumer 会定时查询 Topic 分区信息和消费者信息，来计算自己应该消费的队列
- 除了你提到的这些情况，还可能多个 consumer 的 clientID 一样，然后计算出相同的分区，也会重复，由于 push 默认没有锁的机制，个人认为很难保证同一时间一个队列只有一个消费者消费
- RocketMQ 5.0 支持了 POP 消费，拉取消息时会加锁，因此可以解决这个问题，你可以了解下

**Comment by fsneak** [comment]
我也有这个疑问，如果AllocateMessageQueueStrategy返回的结果重复的话怎么办？
好像RocketMQ至少在4.X好像并没有特别针对这种情况进行处理。
